File: android/titanium/src/java/org/appcelerator/titanium/util/TiResourceUtils.java
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2011 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */
package org.appcelerator.titanium.util;

import org.appcelerator.titanium.ITiAppInfo;
import org.appcelerator.titanium.TiProperties;
import org.appcelerator.titanium.TiApplication;
import org.appcelerator.titanium.io.TiFileFactory;
import org.appcelerator.titanium.io.TiBaseFile;
import org.appcelerator.kroll.util.KrollAssetHelper;
import org.appcelerator.kroll.common.Log;

import java.util.HashMap;
import java.util.regex.Pattern;
import java.util.regex.Matcher;
import android.os.Environment;


public class TiResourceUtils implements KrollAssetHelper.TiResourceUtils
{
	private static final String TAG = "TiResourceUtils";
	
	private static boolean isInit;
	private static boolean useAsset;
	private static boolean useCustomDir;
	private static String sourcePathString;
	private static String baseDirectory;
	private static String pathString;
	
	protected static TiFileFactory tiFileFactory;
	protected static TiApplication tiApp = TiApplication.getInstance();
	protected static TiProperties tiProp =  tiApp.getAppProperties();
	
	public TiResourceUtils()
	{
		
	}
	
	private void init (){
		sourcePathString = tiProp.getString("dataDirectory", "");
		if(sourcePathString != ""){
			useCustomDir = true;
			baseDirectory = getBaseDirectory(sourcePathString);
			pathString = parsePathString(sourcePathString);
		}
		isInit = true;
	}
	
	private boolean isExternalStoragePresent()
	{
		return Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED);
	}
	
	private HashMap getValues(){
		ITiAppInfo appInfo = tiApp.getAppInfo();
		
		HashMap<String, String> values = new HashMap<String, String>();
		values.put("deploytype", appInfo.getDeployType());
		values.put("id", appInfo.getId());
		values.put("publisher", appInfo.getPublisher());
		values.put("url", appInfo.getUrl());
		values.put("name", appInfo.getName());
		values.put("version", appInfo.getVersion());
		values.put("description", appInfo.getDescription());
		values.put("copyright", appInfo.getCopyright());
		values.put("guid", appInfo.getGUID());
		values.put("familyTarget", "android");
		values.put("platform", "android");
		values.put("family", "android");
		values.put("undName", appInfo.getName().replaceAll(" ", "_"));
		
		return values;
	}
	
	private String parsePathString(String path){
		HashMap values = getValues();
		
		Pattern dirPattern = Pattern.compile("(.*Directory):/");
		Pattern valuePattern = Pattern.compile("\\{ *(.*?) *\\}");
		
		Matcher dirRegex = dirPattern.matcher(path);
		
		if(dirRegex.find()){
			path = dirRegex.replaceFirst("");
		}
		
		Matcher valueRegex = valuePattern.matcher(path);
		
		while(valueRegex.find()){
			String matchText = valueRegex.group(0);
			String matchValue = valueRegex.group(1);
			String value = (String) values.get(matchValue);
			if(value != null){
				path = path.replace(matchText, value);
			}
		}
		if(!path.equals("") && !path.endsWith("/")){
			path += "/";
		}
		return path;
	}
	
	private String getBaseDirectory(String path)
	{
		Pattern dirPattern = Pattern.compile("(.*Directory):/");
		Matcher dirRegex = dirPattern.matcher(path);
		
		String basePath = "";
		
		while (dirRegex.find()) {
			String matchText = dirRegex.group(1);
			
			if(matchText.equals("resourcesDirectory")){
				useAsset = true;
				//Use default path
			}else if(matchText.equals("applicationDataDirectory")){
				basePath = tiFileFactory.getDataDirectory(true).toString();
			}else if(matchText.equals("tempDirectory")){
				basePath = tiApp.getTempFileHelper().getTempDirectory().getAbsolutePath();
			}else if(matchText.equals("applicationCacheDirectory")){
				basePath = tiApp.getCacheDir().toString();
			}else if(matchText.equals("externalStorageDirectory") && isExternalStoragePresent()){
				basePath = tiFileFactory.getDataDirectory(false).toString();
			}else{
				useAsset = true;
				//Use default path
			}
		}
		if(!basePath.equals("") && !basePath.endsWith("/")){
			basePath += "/";
		}
		return basePath;
	}
	
	public boolean isExist(String path){
		String basePath = getPath(path);
		if(useAsset) basePath = "app://"+basePath;
		TiBaseFile file = tiFileFactory.createTitaniumFile(basePath, false);
		return file.exists();
	}
	
	public boolean useCustomResourceDirectory()
	{
		if(!isInit){
			init();
		}
		return useCustomDir;
	}
	
	public boolean useAssetDirectory(){
		return useAsset;
	}
	
	public String getBaseDir(){
		return baseDirectory;
	}
	
	public String getBasePath(){
		return baseDirectory+pathString;
	}
	
	public String getPath(String path){
		if(path.startsWith("/")){
			path = path.substring(1);
		}
		return getBasePath()+path;
	}
}

File: iphone/Classes/TiResourceUtils.m
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2013 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiResourceUtils.h"
#import "TiApp.h"
#import "AppModule.h"

@implementation TiResourceUtils

extern NSString * const TI_APPLICATION_DEPLOYTYPE;
extern NSString * const TI_APPLICATION_ID;
extern NSString * const TI_APPLICATION_PUBLISHER;
extern NSString * const TI_APPLICATION_URL;
extern NSString * const TI_APPLICATION_NAME;
extern NSString * const TI_APPLICATION_VERSION;
extern NSString * const TI_APPLICATION_DESCRIPTION;
extern NSString * const TI_APPLICATION_COPYRIGHT;
extern NSString * const TI_APPLICATION_GUID;

static bool useCustomDir = false;
static NSString* sourcePathString;
static NSString* baseDirectory;
static NSString* pathString;

+(void) initialize
{
    sourcePathString = [[TiApp tiAppProperties] objectForKey:@"dataDirectory"];
    if(sourcePathString != nil){
        useCustomDir = true;
        baseDirectory = [[self getBaseDirectory:sourcePathString] retain];
        pathString = [[self parsePathString:sourcePathString] retain];
    }
}

+(NSDictionary*) getValues
{
    NSString* familyTarget = UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad ? @"ipad" : @"iphone";
    NSDictionary* values = [NSDictionary dictionaryWithObjectsAndKeys:
              TI_APPLICATION_DEPLOYTYPE, @"deploytype",
              TI_APPLICATION_ID, @"id",
              TI_APPLICATION_PUBLISHER, @"publisher",
              TI_APPLICATION_URL, @"url",
              TI_APPLICATION_NAME, @"name",
              TI_APPLICATION_VERSION, @"version",
              TI_APPLICATION_DESCRIPTION, @"description",
              TI_APPLICATION_COPYRIGHT, @"copyright",
              TI_APPLICATION_GUID, @"guid",
              familyTarget, @"familyTarget",
              @"ios", @"platform",
              @"iphone", @"family",
              [ TI_APPLICATION_NAME stringByReplacingOccurrencesOfString:@" " withString:@"_"], @"undName", nil];
    return values;
}

+(NSString*) parsePathString:(NSString*) path
{
    NSString* pattern = @"\\{ *(.*?) *\\}";
    NSString* patternDir = @".*Directory:/(.*)";
    NSDictionary* values = [self getValues];

    NSRegularExpression* regexDir = [NSRegularExpression regularExpressionWithPattern: patternDir options:0 error:nil];
    NSTextCheckingResult* matchDir = [[regexDir matchesInString:path options:0 range: NSMakeRange(0, [path length])] firstObject];
    if(matchDir != nil){
        path = [path substringWithRange:[matchDir rangeAtIndex:1]];
    }
    if(![path isEqualToString:@""]){
        NSRegularExpression* regex = [NSRegularExpression regularExpressionWithPattern: pattern options:0 error:nil];
        NSArray* matches = [[[regex matchesInString:path options:0 range: NSMakeRange(0, [path length])] reverseObjectEnumerator] allObjects];
        for (NSTextCheckingResult* match in matches) {
            NSString* matchValue = [path substringWithRange:[match rangeAtIndex:1]];
            NSString* value = values[matchValue];
            if(value != nil){
                path = [path stringByReplacingCharactersInRange:[match range] withString:value];
            }
        }
        if(![path hasSuffix:@"/"]){
            path = [path stringByAppendingString:@"/"];
        }
    }
    return path;
}

+(NSString*) getBaseDirectory:(NSString*) path
{
    NSString* matchText = nil;
    NSRange   searchedRange = NSMakeRange(0, [path length]);
    NSString *pattern = @"(.*Directory):/";
    NSError  *error = nil;
    
    NSRegularExpression* regex = [NSRegularExpression regularExpressionWithPattern: pattern options:0 error:&error];
    NSTextCheckingResult* match = [[regex matchesInString:path options:0 range: searchedRange] firstObject];
    
    NSString* basePath = [[NSBundle mainBundle] resourcePath];
    if(match){
        matchText = [path substringWithRange:[match rangeAtIndex:1]];
        if([matchText isEqualToString:@"resourcesDirectory"]){
            //Use default path
        }else if([matchText isEqualToString:@"applicationDataDirectory"]){
            basePath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject];
        }else if([matchText isEqualToString:@"tempDirectory"]){
            basePath = NSTemporaryDirectory();
        }else if([matchText isEqualToString:@"applicationCacheDirectory"]){
            basePath = [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) firstObject];
        }else if([matchText isEqualToString:@"externalStorageDirectory"]){
            DebugLog(@"[ERROR] 'externalStorageDirectory' supports only android device");
            //Use default path
        }
    }
    if(![basePath hasSuffix:@"/"]){
        basePath = [basePath stringByAppendingString:@"/"];
    }
    return basePath;
}

+(BOOL) useCustomResourceDirectory
{
    return useCustomDir;
}

+(NSString*) getBaseDir
{
    return baseDirectory;
}

+(NSString*) getBasePath
{
    return [baseDirectory stringByAppendingString:pathString];
}

+(NSString*) getPath:(NSString*) path
{
    if(!useCustomDir){
        return nil;
    }
    if([path hasPrefix:@"/"]){
        path = [path substringFromIndex:1];
    }
    NSString* fullPath = [[baseDirectory stringByAppendingString:pathString] stringByAppendingString:path];
    return fullPath;
}

@end
